docs/madr/decisions/011-mesh-traffic-route.md

```diff
@@ -0,0 +1,555 @@
+ to:
+  - targetRef:
+      kind: MeshService
+      name: svc_a
+    default:
+      rules:
+      - matches:
+          path:
+            prefix: /
+        filters:
+        - requestHeader:
+            add:
+            - name: X-Client
+              value:
+```
+
+a user might expect:
+
+```yaml
+spec:
+ targetRef:
+   kind: MeshService
+   name: svc_b
+ to:
+  - targetRef:
+      kind: MeshService
+      name: svc_a
+    default:
+      rules:
+      - matches:
+          path:
+            prefix: /
+        filters:
+        - requestHeader:
+            add:
+            - name: X-Client
+              value:
+        - responseHeader:
+            add:
+            - name: X-Some-Header
+              value: something
+        backendRefs:
+        - weight: 100
+          kind: MeshServiceSubset
+          name: svc_a
+          tags:
+            version: v1
+```
+
+Could we achieve this by always combining `filters` using `matches` as a merge key
+with structural equality and
+specifying `default`/`override` and maybe even `merge` for `filters` _next to_
+`matches`?
+It probably doesn't make sense to merge `backendRefs` so they should be
+limited to `default`/`override`.
+
+Given:
+
+```yaml
+metadata:
+  name: owner
+spec:
+ targetRef:
+   kind: Mesh
+ to:
+  - targetRef:
+      kind: MeshService
+      name: svc_a
+    rules:
+    - matches:
+        path:
+          prefix: /
+      merge:
+        filters:
+        - responseHeader:
+            add:
+            - name: X-Some-Header
+              value: something
+      override:
+        backendRefs:
+        - weight: 90
+          kind: MeshServiceSubset
+          name: svc_a
+          tags:
+            version: v1
+        - weight: 10
+          kind: MeshServiceSubset
+          name: svc_a
+          tags:
+            version: v2
+```
+
+along with:
+
+```yaml
+metadata:
+  name: owner
+spec:
+ targetRef:
+   kind: MeshService
+   name: svc_b
+ to:
+  - targetRef:
+      kind: MeshService
+      name: svc_a
+    rules:
+    - matches:
+        path:
+          prefix: /
+      merge:
+        filters:
+        - requestHeader:
+            add:
+            - name: X-Client
+              value:
+```
+
+would give us the rules:
+
+```yaml
+spec:
+ targetRef:
+   kind: MeshService
+   name: svc_b
+ to:
+  - targetRef:
+      kind: MeshService
+      name: svc_a
+    rules:
+    - matches:
+        path:
+          prefix: /
+      filters:
+      - responseHeader:
+          add:
+          - name: X-Some-Header
+            value: something
+      - requestHeader:
+          add:
+          - name: X-Client
+            value:
+      backendRefs:
+      - weight: 90
+        kind: MeshServiceSubset
+        name: svc_a
+        tags:
+          version: v1
+      - weight: 10
+        kind: MeshServiceSubset
+        name: svc_a
+        tags:
+          version: v2
+```
+
+Both `filters` are applied and the `backendRefs` are taken from the
+less-specific route.
```

Comment:
This covers the addition of a new filter with the same match. What if I want to get rid of the filter?

Reply:
Right, the question is if I have:

> less specific: `merge`

how do I "start over" with a more specific policy? I.e. does this less specific `merge` act more like `override` or `default`.

I certainly fear this language isn't well defined enough to cover all the cases but:

- ```yaml
  ---
  kind: Mesh
  override: [a]
  ---
  kind: MeshService
  merge: [b]
  ```
  gives
  ```yaml
  [a]
  ```
- ```yaml
  ---
  kind: Mesh
  merge: [a]
  ---
  kind: MeshService
  merge: [b]
  ```
  gives
  ```yaml
  [a, b]
  ```
- ```yaml
  ---
  kind: Mesh
  default: [a]
  ---
  kind: MeshService
  merge: [b]
  ```
  gives
  ```yaml
  [b]
  ```

Maybe:

- ```yaml
  ---
  kind: Mesh
  merge: [a]
  ---
  kind: MeshService
  default: [b]
  ```
  gives
  ```yaml
  [b]
  ```

?

Reply:
I'd say it's at the same level as `matches`, i.e. top level of a rule, like suggested in this section but we'd simply disallow `backendRefs` under `merge`.

Reply:
not in question for now